File: app/src/main/java/com/wellness/companion/domain/narrative/MirrorGenerator.kt
package com.wellness.companion.domain.narrative

import com.wellness.companion.data.db.DailyMoodBucket
import com.wellness.companion.data.db.MoodDao
import com.wellness.companion.data.db.NarrativeDao
import com.wellness.companion.data.db.entities.JournalEntry
import com.wellness.companion.data.db.JournalDao
import com.wellness.companion.domain.nlp.TextAnalyzer

/**
 * Generates the monthly "Mirror Moment" — a single-card synthesis of the
 * user's emotional arc, top words, and a highlighted sentence.
 *
 * Designed to run on IO; the UI observes a cached result via StateFlow.
 */
class MirrorGenerator(
    private val moodDao: MoodDao,
    private val journalDao: JournalDao,
    private val narrativeDao: NarrativeDao,
) {
    data class Mirror(
        val periodLabel: String,
        val moodArc: List<DailyMoodBucket>,
        val avgValence: Double,
        val topWords: List<Pair<String, Int>>,
        val highlightTitle: String,
        val highlightSnippet: String,
        val callback: String?,
        val totalEntries: Int,
        val totalMoods: Int,
    )

    suspend fun generate(fromMillis: Long, toMillis: Long, periodLabel: String): Mirror? {
        val entries = narrativeDao.recentEntriesWithKeywords(toMillis, limit = 200)
            .filter { it.createdAt in fromMillis..toMillis }

        if (entries.isEmpty()) return null

        // Mood arc for the sparkline.
        val arc = moodDao.observeDailyAggregateSnapshot(fromMillis, toMillis)
        val avg = moodDao.avgValence(fromMillis, toMillis) ?: 0.0

        // Top words across all entries in the period.
        val allText = entries.joinToString(" ") { it.body }
        val topWords = TextAnalyzer.topWords(allText, limit = 5)

        // Highlight: the longest entry (most emotionally invested).
        val highlight = entries.maxByOrNull { it.body.length }

        // Callback: compare first-week keywords vs last-week keywords.
        val midpoint = fromMillis + (toMillis - fromMillis) / 2
        val firstHalf = entries.filter { it.createdAt < midpoint }
        val secondHalf = entries.filter { it.createdAt >= midpoint }
        val callback = buildCallback(firstHalf, secondHalf)

        return Mirror(
            periodLabel = periodLabel,
            moodArc = arc,
            avgValence = avg,
            topWords = topWords,
            highlightTitle = highlight?.title ?: "",
            highlightSnippet = highlight?.body?.take(160)?.trim() ?: "",
            callback = callback,
            totalEntries = entries.size,
            totalMoods = arc.sumOf { it.sampleCount },
        )
    }

    private fun buildCallback(
        firstHalf: List<com.wellness.companion.data.db.KeywordedEntry>,
        secondHalf: List<com.wellness.companion.data.db.KeywordedEntry>,
    ): String? {
        if (firstHalf.isEmpty() || secondHalf.isEmpty()) return null
        val firstKws = firstHalf.flatMap { it.keywords.split(",") }.filter { it.isNotBlank() }.toSet()
        val secondKws = secondHalf.flatMap { it.keywords.split(",") }.filter { it.isNotBlank() }.toSet()
        val dropped = firstKws - secondKws
        val appeared = secondKws - firstKws
        return when {
            dropped.size >= 2 -> "You started the month writing about ${dropped.take(2).joinToString(" and ")}. By mid-month, you'd moved on."
            appeared.size >= 2 -> "New themes appeared: ${appeared.take(2).joinToString(" and ")}."
            else -> null
        }
    }
}

File: ios/WellnessCompanion/Domain/Narrative/MirrorGenerator.swift
import Foundation

struct Mirror {
    let periodLabel: String
    let moodArc: [DailyMoodBucket]
    let avgValence: Double
    let topWords: [(String, Int)]
    let highlightTitle: String
    let highlightSnippet: String
    let callback: String?
    let totalEntries: Int
    let totalMoods: Int
}

final class MirrorGenerator {
    private let moodStore: MoodStore
    private let journalStore: JournalStore
    private let narrativeStore: NarrativeStore

    init(moodStore: MoodStore, journalStore: JournalStore, narrativeStore: NarrativeStore) {
        self.moodStore = moodStore
        self.journalStore = journalStore
        self.narrativeStore = narrativeStore
    }

    func generate(from: Int64, to: Int64, periodLabel: String) -> Mirror? {
        let entries = narrativeStore.recentEntriesWithKeywords(before: to, limit: 200)
            .filter { $0.createdAt >= from && $0.createdAt <= to }

        guard !entries.isEmpty else { return nil }

        let arc = moodStore.dailyAggregate(from: from, to: to)
        let avg = moodStore.avgValence(from: from, to: to) ?? 0.0

        let allText = entries.map(\.body).joined(separator: " ")
        let topWords = TextAnalyzer.topWords(allText, limit: 5)

        let highlight = entries.max(by: { $0.body.count < $1.body.count })

        let midpoint = from + (to - from) / 2
        let firstHalf = entries.filter { $0.createdAt < midpoint }
        let secondHalf = entries.filter { $0.createdAt >= midpoint }
        let callback = buildCallback(firstHalf: firstHalf, secondHalf: secondHalf)

        return Mirror(
            periodLabel: periodLabel,
            moodArc: arc,
            avgValence: avg,
            topWords: topWords,
            highlightTitle: highlight?.title ?? "",
            highlightSnippet: String((highlight?.body ?? "").prefix(160)).trimmingCharacters(in: .whitespaces),
            callback: callback,
            totalEntries: entries.count,
            totalMoods: arc.reduce(0) { $0 + $1.sampleCount }
        )
    }

    private func buildCallback(firstHalf: [KeywordedEntry], secondHalf: [KeywordedEntry]) -> String? {
        guard !firstHalf.isEmpty, !secondHalf.isEmpty else { return nil }
        let firstKws = Set(firstHalf.flatMap { $0.keywords.split(separator: ",").map(String.init) }.filter { !$0.isEmpty })
        let secondKws = Set(secondHalf.flatMap { $0.keywords.split(separator: ",").map(String.init) }.filter { !$0.isEmpty })
        let dropped = firstKws.subtracting(secondKws)
        let appeared = secondKws.subtracting(firstKws)

        if dropped.count >= 2 {
            return "You started the month writing about \(Array(dropped.prefix(2)).joined(separator: " and ")). By mid-month, you'd moved on."
        } else if appeared.count >= 2 {
            return "New themes appeared: \(Array(appeared.prefix(2)).joined(separator: " and "))."
        }
        return nil
    }
}
